Gradle 4.3-rc-1 breaks spring-boot plugin

A project like the one described in "Steps to Reproduce" worked with Gradle 4.2.1 and earlier.

It does not work with Gradle 4.3-rc-1 anymore.

### Expected Behavior
<!--- If you're describing a bug, tell us what should happen -->
```
BUILD SUCCESSFUL in 0s
```
<!--- If you're suggesting a change/improvement, tell us how it should work -->

### Current Behavior
<!--- If describing a bug, tell us what happens instead of the expected behavior -->
```
FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':bootRepackage'.
> Cannot convert the provided notation to a File or URI: task ':jar'.
  The following types/formats are supported:
    - A String or CharSequence path, for example 'src/main/java' or '/usr/include'.
    - A String or CharSequence URI, for example 'file:/usr/include'.
    - A File instance.
    - A Path instance.
    - A Directory instance.
    - A RegularFile instance.
    - A URI or URL instance.

* Try:
Run with --info or --debug option to get more log output.

* Exception is:
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':bootRepackage'.
        at org.gradle.api.internal.tasks.execution.ValidatingTaskExecuter.execute(ValidatingTaskExecuter.java:50)
        at org.gradle.api.internal.tasks.execution.SkipEmptySourceFilesTaskExecuter.execute(SkipEmptySourceFilesTaskExecuter.java:97)
        at org.gradle.api.internal.tasks.execution.CleanupStaleOutputsExecuter.execute(CleanupStaleOutputsExecuter.java:87)
        at org.gradle.api.internal.tasks.execution.ResolveTaskArtifactStateTaskExecuter.execute(ResolveTaskArtifactStateTaskExecuter.java:52)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:52)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:54)
        at org.gradle.api.internal.tasks.execution.ExecuteAtMostOnceTaskExecuter.execute(ExecuteAtMostOnceTaskExecuter.java:43)
        at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:34)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker$1.run(DefaultTaskGraphExecuter.java:248)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:336)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:328)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:199)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:110)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:241)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:230)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.processTask(DefaultTaskPlanExecutor.java:123)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.access$200(DefaultTaskPlanExecutor.java:79)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:104)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:98)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan.execute(DefaultTaskExecutionPlan.java:626)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan.executeWithTask(DefaultTaskExecutionPlan.java:581)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.run(DefaultTaskPlanExecutor.java:98)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
Caused by: org.gradle.internal.typeconversion.UnsupportedNotationException: Cannot convert the provided notation to a File or URI: task ':jar'.
The following types/formats are supported:
  - A String or CharSequence path, for example 'src/main/java' or '/usr/include'.
  - A String or CharSequence URI, for example 'file:/usr/include'.
  - A File instance.
  - A Path instance.
  - A Directory instance.
  - A RegularFile instance.
  - A URI or URL instance.
        at org.gradle.internal.typeconversion.ErrorHandlingNotationParser.parseNotation(ErrorHandlingNotationParser.java:56)
        at org.gradle.api.internal.file.AbstractFileResolver.convertObjectToFile(AbstractFileResolver.java:126)
        at org.gradle.api.internal.file.AbstractBaseDirFileResolver.doResolve(AbstractBaseDirFileResolver.java:69)
        at org.gradle.api.internal.file.AbstractFileResolver.resolve(AbstractFileResolver.java:86)
        at org.gradle.api.internal.file.AbstractFileResolver.resolve(AbstractFileResolver.java:68)
        at org.gradle.api.internal.tasks.DefaultTaskInputs.toFile(DefaultTaskInputs.java:367)
        at org.gradle.api.internal.tasks.DefaultTaskInputs.access$100(DefaultTaskInputs.java:46)
        at org.gradle.api.internal.tasks.DefaultTaskInputs$6.validate(DefaultTaskInputs.java:345)
        at org.gradle.api.internal.tasks.StaticValue.validate(StaticValue.java:43)
        at org.gradle.api.internal.tasks.DefaultTaskInputFilePropertySpec.validate(DefaultTaskInputFilePropertySpec.java:124)
        at org.gradle.api.internal.tasks.DefaultTaskInputs.validate(DefaultTaskInputs.java:161)
        at org.gradle.api.internal.tasks.execution.ValidatingTaskExecuter.execute(ValidatingTaskExecuter.java:47)
        ... 24 more
```
<!--- If suggesting a change/improvement, explain the difference from current behavior -->

### Steps to Reproduce (for bugs)

1. Create the files below.
1. Execute `gradle wrapper` to use Gradle 4.3-rc-1.
1. Execute `gradlew`.

<!--- Provide a self-contained example project (as an attached archive or a Github project). -->
#### build.gradle
```groovy
buildscript {
    repositories {
        mavenCentral()
    }

    dependencies {
        classpath 'org.springframework.boot:spring-boot-gradle-plugin:1.5.7.RELEASE'
    }
}

wrapper {
    gradleVersion = '4.3-rc-1'
}

apply plugin: 'org.springframework.boot'

bootRepackage {
    // https://github.com/spring-projects/spring-boot/issues/1113
    // https://github.com/spring-projects/spring-boot/issues/8167
    classifier = "repack"
}

defaultTasks 'build'
```

#### src/main/java/foo/Main.java
```java
package foo;

// any class with a public static void main(String[] args) method
public class Main {
    public static void main(String[] args) {
        System.out.println("foo.Main");
    }
}
```

--- a/subprojects/core-api/src/main/java/org/gradle/api/tasks/TaskInputs.java
+++ b/subprojects/core-api/src/main/java/org/gradle/api/tasks/TaskInputs.java
@@ -54,7 +54,7 @@ public interface TaskInputs extends CompatibilityAdapterForTaskInputs {
     /**
      * Registers some input file for this task.
      *
-     * @param path The input file. The given path is evaluated as per {@link org.gradle.api.Project#files(Object...)}.
+     * @param path The input file. The given path is evaluated as per {@link org.gradle.api.Project#file(Object)}.
      * @return a property builder to further configure the property.
      */
     TaskInputFilePropertyBuilder file(Object path);

--- a/subprojects/core/src/integTest/groovy/org/gradle/api/tasks/TaskInputFilePropertiesIntegrationTest.groovy
+++ b/subprojects/core/src/integTest/groovy/org/gradle/api/tasks/TaskInputFilePropertiesIntegrationTest.groovy
@@ -17,6 +17,7 @@
 package org.gradle.api.tasks
 
 import org.gradle.integtests.fixtures.AbstractIntegrationSpec
+import spock.lang.Issue
 import spock.lang.Unroll
 
 class TaskInputFilePropertiesIntegrationTest extends AbstractIntegrationSpec {
@@ -41,5 +42,30 @@ class TaskInputFilePropertiesIntegrationTest extends AbstractIntegrationSpec {
 
         where:
         annotation << [ InputFile, InputDirectory, InputFiles ]
+    }
+
+    @Unroll
+    @Issue("https://github.com/gradle/gradle/issues/3193")
+    def "TaskInputs.#method shows deprecation warning when used with complex input"() {
+        buildFile << """
+            task dependencyTask {
+            }
+
+            task test {
+                inputs.$method(dependencyTask)
+                doFirst {
+                    // Need a task action to not skip this task
+                }
+            }
+        """
+
+        expect:
+        executer.expectDeprecationWarning()
+        succeeds "test"
+
+        output.contains "Using TaskInputs.$method() with something that doesn't resolve to a File object has been deprecated and is scheduled to be removed in Gradle 5.0. Use TaskInputs.files() instead."
+
+        where:
+        method << ["file", "dir"]
     }
 }

--- a/subprojects/core/src/main/java/org/gradle/api/internal/tasks/DefaultTaskInputs.java
+++ b/subprojects/core/src/main/java/org/gradle/api/internal/tasks/DefaultTaskInputs.java
@@ -31,6 +31,8 @@ import org.gradle.api.internal.file.FileTreeInternal;
 import org.gradle.api.internal.file.collections.FileCollectionResolveContext;
 import org.gradle.api.tasks.TaskInputPropertyBuilder;
 import org.gradle.api.tasks.TaskInputs;
+import org.gradle.internal.typeconversion.UnsupportedNotationException;
+import org.gradle.util.DeprecationLogger;
 
 import javax.annotation.Nullable;
 import java.io.File;
@@ -110,7 +112,7 @@ public class DefaultTaskInputs implements TaskInputsInternal {
         return taskMutator.mutate("TaskInputs.file(Object)", new Callable<TaskInputFilePropertyBuilderInternal>() {
             @Override
             public TaskInputFilePropertyBuilderInternal call() {
-                return file(new StaticValue(path));
+                return addSpec(new StaticValue(path), RUNTIME_INPUT_FILE_VALIDATOR);
             }
         });
     }
@@ -125,7 +127,7 @@ public class DefaultTaskInputs implements TaskInputsInternal {
         return taskMutator.mutate("TaskInputs.dir(Object)", new Callable<TaskInputFilePropertyBuilderInternal>() {
             @Override
             public TaskInputFilePropertyBuilderInternal call() {
-                return dir(new StaticValue(dirPath));
+                return addSpec(new StaticValue(dirPath), RUNTIME_INPUT_DIRECTORY_VALIDATOR);
             }
         });
     }
@@ -206,7 +208,7 @@ public class DefaultTaskInputs implements TaskInputsInternal {
     public TaskInputPropertyBuilder property(final String name, @Nullable final Object value) {
         return taskMutator.mutate("TaskInputs.property(String, Object)", new Callable<TaskInputPropertyBuilder>() {
             @Override
-            public TaskInputPropertyBuilder call() throws Exception {
+            public TaskInputPropertyBuilder call() {
                 return property(name, new StaticValue(value));
             }
         });
@@ -362,6 +364,24 @@ public class DefaultTaskInputs implements TaskInputsInternal {
             }
         }
     };
+
+    private static final ValidationAction RUNTIME_INPUT_FILE_VALIDATOR = wrapRuntimeApiValidator("file", INPUT_FILE_VALIDATOR);
+
+    private static final ValidationAction RUNTIME_INPUT_DIRECTORY_VALIDATOR = wrapRuntimeApiValidator("dir", INPUT_DIRECTORY_VALIDATOR);
+
+    private static ValidationAction wrapRuntimeApiValidator(final String method, final ValidationAction validator) {
+        return new ValidationAction() {
+            @Override
+            public void validate(String propertyName, Object value, TaskValidationContext context, TaskValidationContext.Severity severity) {
+                try {
+                    validator.validate(propertyName, value, context, severity);
+                } catch (UnsupportedNotationException ex) {
+                    DeprecationLogger.nagUserOfDeprecated("Using TaskInputs." + method + "() with something that doesn't resolve to a File object", "Use TaskInputs.files() instead");
+                }
+            }
+        };
+    }
+
 
     private static File toFile(TaskValidationContext context, Object value) {
         return context.getResolver().resolve(value);

--- a/subprojects/docs/src/docs/release/notes.md
+++ b/subprojects/docs/src/docs/release/notes.md
@@ -305,6 +305,26 @@ task myTask {
 }
 ```
 
+### Calling `TaskInputs` methods with incorrect parameters
+
+`TaskInputs.file()` used to accept anything that `files()` would, and the same was true for `dir()`. Starting with Gradle 4.3 this behavior is deprecated, and only values that resolve to a single file or directory are accepted.
+
+Do not do this:
+
+```
+task myTask {
+    inputs.dir fileTree(...)
+}
+```
+
+Do this instead:
+
+```
+task myTask {
+    inputs.files fileTree(...)
+}
+```
+
 ### Deprecation of `TaskInternal.execute()`
 
 In this release we deprecate calling `TaskInternal.execute()`. Calling `task.execute()` should never be necessary.

